fdbserver/TLogServer.actor.cpp

```diff
@@ -1483,21 +1511,141 @@ ACTOR Future<Void> tLogCommit(
 		return Void();
 	}
 
+    state Version execVersion = invalidVersion;
+    state ExecCmdValueString execArg();
+    state TLogQueueEntryRef qe;
+    state StringRef execCmd;
+    state StringRef param2;
+	state vector<Tag> execTags;
+
 	if (logData->version.get() == req.prevVersion) {  // Not a duplicate (check relies on no waiting between here and self->version.set() below!)
 		if(req.debugID.present())
 			g_traceBatch.addEvent("CommitDebug", tlogDebugID.get().first(), "TLog.tLogCommit.Before");
 
-		//TraceEvent("TLogCommit", logData->logId).detail("Version", req.version);
-		commitMessages(self, logData, req.version, req.arena, req.messages);
-
-		logData->knownCommittedVersion = std::max(logData->knownCommittedVersion, req.knownCommittedVersion);
-
 		// Log the changes to the persistent queue, to be committed by commitQueue()
-		TLogQueueEntryRef qe;
 		qe.version = req.version;
 		qe.knownCommittedVersion = logData->knownCommittedVersion;
 		qe.messages = req.messages;
 		qe.id = logData->logId;
+
+		if (req.hasExecOp) {
+			// inspect the messages to find if there is an Exec type and print
+            // it. message are prefixed by the length of the message and each
+            // field is prefixed by the length too
+			uint8_t type = MutationRef::MAX_ATOMIC_OP;
+			{
+				ArenaReader rd(req.arena, qe.messages, Unversioned());
+				int32_t messageLength, rawLength;
+				uint16_t tagCount;
+				uint32_t sub;
+				while(!rd.empty()) {
+					Tag tmpTag;
+					rd.checkpoint();
+					rd >> messageLength >> sub >> tagCount;
+					for(int i = 0; i < tagCount; i++) {
+						rd >> tmpTag;
+						execTags.push_back(tmpTag);
+					}
+					rd >> type;
+					if (type == MutationRef::Exec) {
+						break;
+					}
+					rawLength = messageLength + sizeof(messageLength);
+					rd.rewind();
+					rd.readBytes(rawLength);
+				}
+				int32_t len = 0;
+				if (type == MutationRef::Exec) {
```

Comment:
You should just skip parsing any mutations that have the txsTag on them, as they aren't mutations and don't follow the normal mutation format.  I think this works by accident, because MutationRef::Exec is greater than all of the KeyValueStoreMemory::OpType enums.

I think you're also going to need to add a `wait(yield())` to this loop to avoid a SlowTask on large commits.

Overall, this continues to make me still uncertain of the advantages of parsing the Exec mutation back out of the stream, rather than just duplicating it as a new part of a TLogCommitRequest.

Reply:
> You should just skip parsing any mutations that have the txsTag on them, as they aren't mutations and don't follow the normal mutation format. I think this works by accident, because MutationRef::Exec is greater than all of the KeyValueStoreMemory::OpType enums.

Sounds good.

> I think you're also going to need to add a wait(yield()) to this loop to avoid a SlowTask on large commits.

FIRST_IN_BATCH optimization will solve the problem, I am making that change in a following commit.

> Overall, this continues to make me still uncertain of the advantages of parsing the Exec mutation back out of the stream, rather than just duplicating it as a new part of a TLogCommitRequest.

If we want to pass the exec mutation in the future to remote region and have them snapshotted there then we need to parse the stream, because logRouter and remote TLog will pull the mutations as against getting a commit request. I think it is acceptable and cheap if the exec op with the FIRST_IN_BATCH optimization.